java/src/main/java/com/google/oak/util/Result.java

```diff
@@ -0,0 +1,148 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package com.google.oak.util;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Class {@code Result} is a generic class representing the result of an
+ * operation that could either succeed and produce an instance of {@code R}, or
+ * fail and produce an error of type {@code E}. An instance of {@code Result}
+ * can encapsulate both outcomes.
+ *
+ * <p>The main invariant that instances of this class maintain is that each object always either
+ * contains a non-null success value or a non-null error value.
+ */
+public class Result<R, E> {
```

Comment:
I like this idea. Let me try it and see if I can make it work.

Reply:
I tried a little bit, but I am not sure if it will work. In particular, IIUC you cannot have a generic `enum` in java. I'd be happy to look into it together at some point :)

Reply:
Unfortunately Java does not have "real" algebraic data types (in particular, discriminated unions), so this is not possible. :disappointed: